File: backend/app/models/protocol.py
"""
Protocol domain models.
Mirrors the JSON Schema in shared/schemas/protocol-schema.json.
"""
from enum import Enum

from pydantic import BaseModel, Field


class ProviderLevel(str, Enum):
    """EMS Provider certification levels."""
    ALL = "ALL"           # BLS + ALS
    BLS = "BLS"           # Basic Life Support
    ALS = "ALS"           # Advanced Life Support (includes BLS scope)
    PARAMEDIC = "PARAMEDIC"  # Paramedic-only interventions
    OLMC = "OLMC"         # Requires Medical Control order


class PatientType(str, Enum):
    """Patient age categories per PWC PCM definitions."""
    ADULT = "adult"       # >= 18 years per PWC definition
    PEDIATRIC = "pediatric"  # < 18 years
    NEONATE = "neonate"   # Birth to 1 month
    ALL = "all"


class Route(str, Enum):
    """Medication administration routes."""
    PO = "PO"      # Per Os (oral)
    SL = "SL"      # Sublingual
    IV = "IV"      # Intravenous
    IO = "IO"      # Intraosseous
    IM = "IM"      # Intramuscular
    IN = "IN"      # Intranasal
    NEB = "NEB"    # Nebulized
    ODT = "ODT"    # Oral Dissolving Tablet
    TOPICAL = "TOPICAL"


class VitalSignCriteria(BaseModel):
    """Vital sign thresholds for protocol matching."""
    sbp_below: int | None = None
    sbp_above: int | None = None
    map_below: int | None = None
    hr_below: int | None = None
    hr_above: int | None = None
    spo2_below: int | None = None
    gcs_below: int | None = None
    etco2_below: int | None = None
    etco2_above: int | None = None


class Criteria(BaseModel):
    """Inclusion/exclusion criteria for protocol matching."""
    age_min_months: int | None = None
    age_max_months: int | None = None
    symptoms: list[str] = Field(default_factory=list)
    vital_signs: VitalSignCriteria | None = None
    exclusions: list[str] = Field(default_factory=list)


class MedicationDose(BaseModel):
    """Specific medication dosing within a step."""
    medication_id: str
    dose: str
    dose_per_kg: float | None = None
    max_dose: str | None = None
    route: Route
    rate: str | None = None


class Step(BaseModel):
    """Individual treatment step within a protocol."""
    order: int
    provider_level: ProviderLevel
    action: str
    medication: MedicationDose | None = None
    condition: str | None = None  # e.g., "SBP > 100"
    contraindications: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
    repeat: str | None = None  # e.g., "Every 5 minutes x3"
    cross_reference: str | None = None  # Protocol/Procedure ID


class Protocol(BaseModel):
    """
    Complete EMS Protocol definition.
    Maps to PWC PCM structure: All Providers -> ALS -> Paramedic -> Medical Control
    """
    id: str
    title: str
    category: str
    patient_type: PatientType
    tags: list[str] = Field(default_factory=list)
    criteria: Criteria | None = None
    steps: list[Step]
    cross_references: list[str] = Field(default_factory=list)

File: frontend/src/types/protocol.ts
/**
 * Protocol type definitions.
 * These MUST mirror backend/app/models/ exactly.
 * Source of truth: shared/schemas/protocol-schema.json
 */

export type ProviderLevel = 'ALL' | 'BLS' | 'ALS' | 'PARAMEDIC' | 'OLMC';
export type PatientType = 'adult' | 'pediatric' | 'neonate' | 'all';
export type Route = 'PO' | 'SL' | 'IV' | 'IO' | 'IM' | 'IN' | 'NEB' | 'ODT' | 'TOPICAL';

export interface MedicalControlContact {
  facility: string;
  phone: string;
  radio_channel?: string;
}

export interface MedicalControl {
  contacts: MedicalControlContact[];
  poison_control: string;
  chemtrec: string;
}

export interface JurisdictionMeta {
  jurisdiction: string;
  version: string;
  effective_date: string;
  revision_date?: string;
  medical_control: MedicalControl;
}

export interface VitalSignCriteria {
  sbp_below?: number;
  sbp_above?: number;
  map_below?: number;
  hr_below?: number;
  hr_above?: number;
  spo2_below?: number;
  gcs_below?: number;
  etco2_below?: number;
  etco2_above?: number;
}

export interface Criteria {
  age_min_months?: number;
  age_max_months?: number;
  symptoms?: string[];
  vital_signs?: VitalSignCriteria;
  exclusions?: string[];
}

export interface MedicationDose {
  medication_id: string;
  dose: string;
  dose_per_kg?: number;
  max_dose?: string;
  route: Route;
  rate?: string;
}

export interface Step {
  order: number;
  provider_level: ProviderLevel;
  action: string;
  medication?: MedicationDose;
  condition?: string;
  contraindications?: string[];
  notes?: string[];
  repeat?: string;
  cross_reference?: string;
}

export interface Protocol {
  id: string;
  title: string;
  category: string;
  patient_type: PatientType;
  tags: string[];
  criteria?: Criteria;
  steps: Step[];
  cross_references?: string[];
}

export interface DosingInfo {
  indication: string;
  dose: string;
  route: string;
  max_dose?: string;
  notes?: string;
}

export interface Medication {
  id: string;
  generic_name: string;
  trade_names: string[];
  drug_class: string;
  mechanism?: string;
  indications: string[];
  contraindications: string[];
  adult_dosing: DosingInfo[];
  pediatric_dosing: DosingInfo[];
}

export interface Procedure {
  id: string;
  title: string;
  procedure_type: string;
  authorized_personnel: string[];
  indications: string[];
  contraindications: string[];
  equipment: string[];
  steps: string[];
}

export interface ProtocolBundle {
  meta: JurisdictionMeta;
  protocols: Protocol[];
  medications: Medication[];
  procedures: Procedure[];
}
